guests/eth-block/src/main.rs

```diff
@@ -36,8 +76,50 @@ pub fn main() {
         .expect("Failed to execute transactions")
         .build::<BuildFromMemDbStrategy>()
         .expect("Failed to build the resulting block");
-    // Output the resulting block's hash to the journal
-    env::commit(&output.hash());
-    // Leak memory, save cycles
-    core::mem::forget(output);
+
+    // get hash of the block header and print it
+
+    let output_hash = output.hash();
+    let output_hash_str = output_hash.to_string();
+    println!("{}", output_hash_str);
+
+    // generate random Ethereum public & private keys
+
+    use secp256k1::rand::rngs::OsRng;
+    use secp256k1::{Secp256k1, Message};
+    use secp256k1::hashes::sha256;
+
+    let secp = Secp256k1::new();
+    let (secret_key, public_key) = secp.generate_keypair(&mut OsRng);
+
+    // sign the above hash of the block header with the above secret key
+
+    let message = Message::from_hashed_data::<sha256::Hash>(output_hash_str.as_bytes());
+    let sig = secp.sign_ecdsa(&message, &secret_key);
+    // eprintln!("Private key: {}", secret_key.);
+    eprintln!("Public key: 0x{}", public_key);
+    eprintln!("Signature: 0x{}", sig);
```

Comment:
What do you think is the best way of getting data out (and data in) of an SGX program? I guess the Automata guy said all communication via TLS, but I don't know if you have a different idea.

Reply:
Yeah, using RA-TLS the alternative, higher abstraction level way to do attestation (see 3 level of abstractions of attestation nicely summarized [here](https://gramine.readthedocs.io/en/stable/attestation.html#attestation-and-secret-provisioning)). But in that case we would need to have server-like, long running process serving TLS requests (instead of one-shot/run-and-exit application).

[RA-TLS is mid-level interface](https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface) (meaning not the highest level and not the lowest level attestation abstraction). The essence from [the above mentioned link](https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface):
> [RA-TLS](https://gramine.readthedocs.io/en/stable/glossary.html#term-ra-tls) interface hides the complexity of the low-level /dev/attestation flows and provides a simple and powerful abstraction of a TLS connection between the TEE and the remote user (enhanced with remote-attestation flows). Using RA-TLS, the application can securely send and receive arbitrary data to/from the remote user. RA-TLS is currently tied to Intel SGX but can be adapted for other TEEs.
>
> RA-TLS integrates Intel SGX remote attestation into the TLS connection setup. Conceptually, it extends the standard X.509 certificate with SGX-related information (SGX quote). The additional information allows the remote user (verifier) of the certificate to verify that it is indeed communicating with an SGX enclave (attester).

I thought we may go with the [low level way of doing SGX attestation](https://gramine.readthedocs.io/en/stable/attestation.html#low-level-dev-attestation-interface) where the app running in SGX enclave fetches so called [quote](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-remote-attestation) that can be later verified (potentially in the smart contract in the future):
> In remote attestation, the attesting SGX enclave collects attestation evidence in the form of an [SGX Quote](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-sgx-quote) using the [Quoting Enclave](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-quoting-enclave) (and the [Provisioning Enclave](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-provisioning-enclave) if required). This form of attestation is used to send the attestation evidence to a remote party (not on the same physical machine).

From https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-sgx-quote:
>The SGX quote is the proof of trustworthiness of the enclave and is used during [Remote Attestation](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-remote-attestation). The attesting enclave generates the enclave-specific [SGX Report](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-sgx-report), sends the request to the [Quoting Enclave](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-quoting-enclave) using [Local Attestation](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-local-attestation), and the Quoting Enclave returns back the SGX quote with the SGX report embedded in it. The resulting SGX quote contains the enclave’s measurement, attributes and other security-relevant fields, and is tied to the identity of the [Quoting Enclave](https://gramine.readthedocs.io/en/stable/sgx-intro.html#term-quoting-enclave) to prove its authenticity. **The obtained SGX quote may be later sent to the verifying remote party, which examines the SGX quote and gains trust in the remote enclave**.